### PR TITLE
test CUDA

### PR DESCRIPTION
**Pull request by loriab**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] ~Reset the build number to `0` (if the version changed)~
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

~I'm concerned that CUDA isn't actually being activated and that #58 passing isn't testing what we think. This tries to activate more explicitly to serve as comparison. `linux-64` only for CUDA.~

- [x] enable cuda for cuda builds :-) and bump derivatives down to second to avoid abandoned builds. (we could also try restricting the number of ninja threads to free up memory for the higher derivatives.)
- [x] suppress CI testing of cuda builds, which are demonstrated to fail. thanks to @njzjz and @caic99, downloaded cuda artifacts are now known to operate.

### CODE REVIEW

**Comment by loriab** [comment]
The problem seems to be that the operational check fails when cuda is enabled `Segmentation fault      (core dumped) xc-info 100` . https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=795193&view=logs&j=4f922444-fdfe-5dcf-b824-02f86439ef14&t=937c195f-508d-5135-dc9f-d4b5730df0f7&l=2671

**Comment by loriab** [comment]
@awvwgk , would you mind confirming what worked at time of your initial CUDA addition? Did `xc-info 100` work then? We merged the C/Fortran/Py split the CUDA PRs so near that I suspect something got lost, especially `ENABLE_CUDA=ON`.

**Comment by jakirkham** [comment]
@awvwgk , would you have a moment to follow up here? We would appreciate hearing your insights 🙂

**Comment by awvwgk** [comment]
I did run the build of the recipe locally on a GPU enabled machine. I can check whether I still have the artifact to check again, but as far as I remember it was working as intended.

**Comment by jakirkham** [comment]
@conda-forge-admin , please re-render

**Comment by jakirkham** [comment]
CI logs had expired. So tried to rerun, but it looks like there are conflicts

A question about the test. Is this trying to use a GPU as part of the test? If so, there are not GPUs to test with on CI. So it may be expected that it doesn't work

We could configure CI to cache a package from the PR so it could be downloaded for testing (though we might need to disable the problematic test to get far enough in the CI build to create a package). Maybe then someone can explore whether that package works as expected on a system with a GPU (or share what issues they uncover)

**Comment by loriab** [comment]
@conda-forge-admin , please re-render

**Comment by loriab** [comment]
> A question about the test. Is this trying to use a GPU as part of the test? If so, there are not GPUs to test with on CI. So it may be expected that it doesn't work

The offending line was `xc-info 100`. That's printing a few lines of stored text, so no intended GPU involvement. But `xc-info` is a compiled executable that might have GPU libs linked in, so maybe it should be an xfail.

> We could configure CI to cache a package from the PR so it could be downloaded for testing (though we might need to disable the problematic test to get far enough in the CI build to create a package). Maybe then someone can explore whether that package works as expected on a system with a GPU (or share what issues they uncover)

Good notion. I've disabled the `xc-info` for the gpu-build and asked it to save artifacts. I may need to disable more pylibxc stuff to get a passing build. Unfortunately I don't have a local GPU box for trial.

**Comment by jakirkham** [comment]
@awvwgk could you please help us test out the binaries here?

Here are [some relevant docs on using CI artifacts locally]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#downloading-prebuilt-packages-from-ci )

Some examples with screenshots in issue: https://github.com/conda-forge/conda-forge.github.io/issues/1424

**Comment by jakirkham** [comment]
@njzjz , as `abacus` is [a downstream user of `libxc`]( https://github.com/conda-forge/abacus-feedstock/blob/24e07132d01c4c6595ef32eff66398d4e63506dc/recipe/meta.yaml#L82-L83 ), am wondering if you would be willing to help us test the CI package artifacts

**Comment by njzjz** [comment]
> @njzjz , as `abacus` is [a downstream user of `libxc`](https://github.com/conda-forge/abacus-feedstock/blob/24e07132d01c4c6595ef32eff66398d4e63506dc/recipe/meta.yaml#L82-L83), am wondering if you would be willing to help us test the CI package artifacts

OK, I will take a test.
cc @caic99

**Comment by caic99** [comment]
Hi @loriab ,
I tested the libxc cuda build ([hc895d526af](https://artprodeus21.artifacts.visualstudio.com/A910fa339-c7c2-46e8-a579-7ea247548706/84710dde-1620-425b-80d0-4cf5baca359d/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2NvbmRhLWZvcmdlL3Byb2plY3RJZC84NDcxMGRkZS0xNjIwLTQyNWItODBkMC00Y2Y1YmFjYTM1OWQvYnVpbGRJZC84Mjk2OTgvYXJ0aWZhY3ROYW1lL2NvbmRhX2FydGlmYWN0c18yMDIzMTEyMy4zLjFfbGludXhfNjRfY19jb21waWxlcl92ZXJzaW9uMTJjdWRhX2NfaGM4OTVkNTI2YWY1/content?format=zip)) with ABACUS. Unfortunately, it produces an error:
`error: output variable, out->v2rhotau, is a null pointer`
The default recipe of ABACUS uses `conda-forge::libxc-c-6.2.2-cuda_h60e1ed8_2`. Is that version GPU accelerated? Do we need to adapt codes for the new package?

[Test case used](https://github.com/deepmodeling/abacus-develop/tree/develop/tests/libxc/Si_gammapoint_nspin1)

**Comment by njzjz** [comment]
I read libxc documentation, and I'm afraid that abacus doesn't support libxc cuda. It says

> When running with libxc compiled with Cuda, both the input and output arrays must always be allocated on the GPU (or using unified memory). Libxc will fail (most likely you will get a segmentation fault) if a CPU array is passed.

@caic99 So we may need to force the libxc CPU version before abacus really supports it. If the previous libxc cuda package works, it may not actually be compiled with cuda.

**Comment by loriab** [comment]
Thank you for testing, @caic99 . I can't look at details for the next few days, I'm afraid (conference).

The `libxc-c-6.2.2-cuda_h60e1ed8_2` you're using by default nominally _should_ be GPU accelerated. The kerfluffle of this test PR is that it seems likely that the libxc-c-*-cuda packages currently on c-f aren't actually conda enabled (not the libxc+cuda+recipe author's fault -- there was a lot of recipe changes around then).

**Comment by jakirkham** [comment]
Thanks Jinzhe and Chun! 🙏

Really appreciate the help testing 🙂

**Comment by caic99** [comment]
> Edit: Are either of you able to run `xc-info 100` with the package?

@jakirkham It works

```
xc-info 100
libxc version 6.2.2
S. Lehtola, C. Steigemann, M. J.T. Oliveira, and M. A.L. Marques.,  SoftwareX 7, 1–5 (2018)
doi: 10.1016/j.softx.2017.11.002

   func_id: 100
      name: gga_c_tca
    family: XC_FAMILY_GGA
      kind: XC_CORRELATION
   comment: Tognetti, Cortona, Adamo

Reference(s):
  *) V. Tognetti, P. Cortona, and C. Adamo.,  J. Chem. Phys. 128, 034101 (2008)
     doi: 10.1063/1.2816137
     bibtex key: Tognetti2008_034101

Implementation has support for:
  *) energy
  *) first derivative
  *) second derivative

Default thresholds:
density: 1.000000e-14
   zeta: 2.220446e-16
  sigma: 2.154435e-19

Functional has no external parameters.
```

File: conda-forge.yml [hunk 1]
```diff
@@ -1,3 +1,5 @@
+azure:
+  store_build_artifacts: true
```
Comment (jakirkham) [inline conda-forge.yml:2]: Guessing this is just for testing

Whenever testing is done, would recommend dropping this and re-rendering

```suggestion
```

**Comment by jakirkham** [comment]
Thanks Chun! 🙏

**Comment by jakirkham** [comment]
Going back to question about libraries linked into `xc-info`, there is some info in the [CI log]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=829698&view=logs&j=4f922444-fdfe-5dcf-b824-02f86439ef14&t=937c195f-508d-5135-dc9f-d4b5730df0f7&l=2625 )

Looking at the results there don't appear to be any dynamic CUDA libraries linked. However there could be static CUDA libraries that are linked in. This is important to note as `cudart` is statically linked by default (NVCC and CMake default)

```
+ ldd -v $PREFIX/bin/xc-info
	linux-vdso.so.1 =>  (0x00007ffd8f7f7000)
	libxc.so.12 => $PREFIX/bin/../lib/libxc.so.12 (0x00007f3c1d600000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f3c1d3f8000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f3c1d1dc000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f3c1cfd8000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f3c1cc0a000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f3c2b2e9000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f3c1c908000)
```

**Comment by caic99** [comment]
> Do we need to adapt codes for the new package?

[Libxc docs](https://gitlab.com/libxc/libxc#gpu-support-with-cuda) said
> When running with libxc compiled with Cuda, both the input and output arrays must always be allocated on the GPU (or using unified memory).
Libxc will fail (most likely you will get a segmentation fault) if a CPU array is passed. 

Does it mean ABACUS should change the array allocation to GPU if CUDA-supported Libxc is used? If so, how to determine if the Libxc provided is CUDA-supported on runtime/compile time?

**Comment by jakirkham** [comment]
^ Am guessing this is a question for @njzjz

**Comment by njzjz** [comment]
@caic99, I submitted https://github.com/conda-forge/abacus-feedstock/pull/20 since we should use the CPU version of libxc for the current codes. For the GPU interface, perhaps we should create a new issue in the abacus repository.

**Comment by jakirkham** [comment]
@loriab , hope the conference well

Not sure if you saw, but it looks like testing `libxc` with CUDA [works]( https://github.com/conda-forge/libxc-feedstock/pull/59#issuecomment-1844815261 )

Think the segfault is just due to lack of a working GPU. So maybe skipping the tests when building for CUDA is appropriate

**Comment by loriab** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml [hunk 2]
```diff
@@ -17,7 +17,7 @@ source:
   sha256: {{ sha256 }}
 
 build:
-  number: 2
+  number: 3
   skip: true  # [not linux64 and cuda_compiler_version != "None"]
   skip: true  # [linux64 and cuda_compiler_version == "10.2"]
```
Comment (jakirkham) [inline recipe/meta.yaml:22]: Think we can drop this one as conda-forge now only supports CUDA 11.2+ ( https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/1708 )

```suggestion
```

**Comment by loriab** [comment]
> Not sure if you saw, but it looks like testing libxc with CUDA https://github.com/conda-forge/libxc-feedstock/pull/59#issuecomment-1844815261

> Think the segfault is just due to lack of a working GPU. So maybe skipping the tests when building for CUDA is appropriate

thanks, @jakirkham and testers. I've set this PR up to be mergeable and will see if @jan-janssen wants to look it over; otherwise will merge tomorrow. Then, I suppose, back to #58.

**Comment by jakirkham** [comment]
Thanks everyone! 🙏

**Pull request merged**